sklearn_numba_dpex/kmeans/engine.py

```diff
@@ -156,155 +205,170 @@ def kmeans_single(self, X, sample_weight, centers_init):
             inertia,
             # XXX: having a C-contiguous centroid array is expected in sklearn in some
             # unit test and by the cython engine.
-            np.ascontiguousarray(
-                dpt.asnumpy(best_centroids).astype(output_dtype, copy=False)
-            ),
+            # ???: rather that returning whatever dtype the driver returns (which might
+            # depends on device support for float64), shouldn't we cast to a dtype that
+            # is always consistent with the input ? (e.g. cast to float64 if the input
+            # was given as float64 ?) But what assumptions can we make on the input
+            # so we can infer its input dtype without risking triggering a copy of it ?
+            np.ascontiguousarray(dpt.asnumpy(best_centroids.T)),
             n_iteration,
         )
 
+    def prepare_prediction(self, X, sample_weight):
+        X = self._validate_data(X, reset=False)
+        sample_weight = self._check_sample_weight(sample_weight, X)
+        return X, sample_weight
+
     def get_labels(self, X, sample_weight):
-        labels, _ = self._get_labels_inertia(X, with_inertia=False)
+        # TODO: sample_weight actually not used for get_labels. Fix in sklearn ?
```

Comment:
Is there an issue already open for it?

Reply:
Done in https://github.com/scikit-learn/scikit-learn/issues/25066